client/src/main/java/com/alibaba/nacos/client/config/impl/ClientWorker.java

```diff
@@ -190,39 +181,27 @@ void removeCache(String dataId, String group, String tenant) {
      * @return cache data
      */
     public CacheData addCacheDataIfAbsent(String dataId, String group) {
-        CacheData cache = getCache(dataId, group);
-        if (null != cache) {
-            return cache;
-        }
-        
-        String key = GroupKey.getKey(dataId, group);
-        cache = new CacheData(configFilterChainManager, agent.getName(), dataId, group);
-        
-        synchronized (cacheMap) {
-            CacheData cacheFromMap = getCache(dataId, group);
-            // multiple listeners on the same dataid+group and race condition,so double check again
-            //other listener thread beat me to set to cacheMap
-            if (null != cacheFromMap) {
-                cache = cacheFromMap;
-                //reset so that server not hang this check
-                cache.setInitializing(true);
-            } else {
-                int taskId = cacheMap.get().size() / (int) ParamUtil.getPerTaskConfigSize();
-                cache.setTaskId(taskId);
-            }
-            
-            Map<String, CacheData> copy = new HashMap<String, CacheData>(cacheMap.get());
-            copy.put(key, cache);
-            cacheMap.set(copy);
+        String key =  GroupKey.getKey(dataId, group);
+
+        CacheData cacheData = new CacheData(configFilterChainManager, agent.getName(), dataId, group);
+        // multiple listeners on the same dataid+group and race condition
+        CacheData lastCacheData = cacheMap.putIfAbsent(key, cacheData);
+
+        if (lastCacheData == null) {
+            int taskId = cacheMap.size() / (int) ParamUtil.getPerTaskConfigSize();
+            lastCacheData = cacheData;
+            lastCacheData.setTaskId(taskId);
         }
-        
+        // reset so that server not hang this check
+        lastCacheData.setInitializing(true);
```

Comment:
if lastCacheData == null 

you can't set CacheData.Initializing =true

Reply:
lastCacheData not null
```
        **CacheData cacheData = new CacheData(configFilterChainManager, agent.getName(), dataId, group, tenant);**
if (lastCacheData == null) {
            //fix issue # 1317
            if (enableRemoteSyncConfig) {
                String[] ct = getServerConfig(dataId, group, tenant, 3000L);
                cacheData.setContent(ct[0]);
            }
            **lastCacheData = cacheData;**
        }
``

Reply:
There are some different with old logic.
In old logic, it seems that only not new cacheData need to set Initializing.
But in your new logic, all cacheData what ever old or new will set Initializing. It might cause bugs.

```
CacheData cacheFromMap = getCache(dataId, group);
            // multiple listeners on the same dataid+group and race condition,so double check again
            //other listener thread beat me to set to cacheMap
            if (null != cacheFromMap) { // the cache data from cache, means it is old cache
                cache = cacheFromMap; // set cache as old cache
                //reset so that server not hang this check
                cache.setInitializing(true); // set old cache Initializing
            } else {
                int taskId = cacheMap.get().size() / (int) ParamUtil.getPerTaskConfigSize();
                cache.setTaskId(taskId);
            }

```

```
CacheData cacheData = new CacheData(configFilterChainManager, agent.getName(), dataId, group);
CacheData lastCacheData = cacheMap.putIfAbsent(key, cacheData);
if (lastCacheData == null) {   // old cache non-exist
    ...
    lastCacheData = cacheData; // set old cache as new cache
    ...
}
lastCacheData.setInitializing(true); // set new cache Initializing, 
```